### DISCUSSION TITLE
Random grabbing of separate .pngs to create a layered image, finally merge to flat image output?

### DISCUSSION THREAD

**Discussion by DaddyKooL18** [ImageMagick/ImageMagick#3815] [Help]
I am trying to create random avatars ... the following is a good example of the kind of thing ...

https://mobile.twitter.com/ArabianCamels/status/1397674016064688129

I wondered if ImageMagick could grab first a random background from the background folder, then a random face from faces folder, layer it on top of background ... then the same for Eyes ... etc. etc.

then output the new random merged result sequentially .... then repeat up to X amount of times

.....

Dir Structure for example

\Background\1.png
\Background\2.png
\Background\3.png
etc.

\Faces\1.png
\Faces\2.png
\Faces\3.png

\Eyes\1.png
\Eyes\2.png
etc. etc.
 

\Output\image001.png
\Output\image002.png

etc.

I have got as far as:- 

magick composite .... Then got really confused :(

Do you think there is a commend line or batch that would do this please?

Thank you for any pointers .. My brain has almost exploded looking for a solution the last couple of days! :P

**Comment by fmw42** [comment]
What platform/OS?  What version of ImageMagic? ImageMagick can do that but you have to script it. Scripting is different for Windows than Unix. Also may be different for IM 7 and IM 6.

**Reply by DaddyKooL18** [reply]
Hi there! .. Thank you for the rapid response! :)

ImageMagick-7.0.11-Q16-HDRI    - I am on Windows 10  ... and NOT a programmer by any stretch of the imagination :(

**Reply by fmw42** [reply]
Are you on Windows 10 Unix or plain Windows (DOS)?  If the latter, one of the Windows users will have to help further. I only work on Mac. As I said above it will need scripting to grab the files from different folders. Then one can use

`magick background foreground  -geometry +X+Y -compose over -composite background`

for each foreground object to place it at X,Y on the background. The new background would be used for the next foreground object in the next step. Note that background and foreground are placeholders for your actual file names.

You need to script the part where you look in each directory and select a random file. That is scripting your operating system to do the file selection and retrieval.  The OS can compute a random number or ImageMagick can do that.

**Reply by DaddyKooL18** [reply]
Thank you :)

Yes that command makes sense :) .. so yes I just need the Random selection part (If I can have more than two 'adds')

:)

Let's see if some Windows people are reading haha :)

**Reply by fmw42** [reply]
If you want to composite multiple objects at once, you can use -flatten as follows:

`magick background ( -page +X1+Y1 object1 ) ( -page +X2+Y2 object2 ) ... ( -page +Xn+Yn objectn ) -flatten result`

See https://legacy.imagemagick.org/Usage/layers/#flatten

**Comment by snibgo** [comment]
Yeah, I use Windows.

First, decide on what magick command does what you need to build an output from one background, one face, one eyes, and so on. I suggest using "magick" alone, not "magick composite". If you show sample face and eyes etc, we can give better advice.

A Windows BAT script could be something like this:
```
set NumBack=9
set NumFace=5
set NumEyes=12

set FMT=^
fBack=Background\\%%[fx:floor(rand()*%NumBack%)+1].png\n^
fFace=Faces\\%%[fx:floor(rand()*%NumFace%)+1].png\n^
fEyes=Eyes\\%%[fx:floor(rand()*%NumEyes%)+1].png\n

for /F "usebackq" %%L in (`%IMG7%magick identify ^
  -format "%FMT%" ^
  xc:`) do set %%L

rem Then insert a magick command that composites files %fEyes% and %fFace% over %fBack%.
```

**Reply by DaddyKooL18** [reply]
Thank you too!

That looks amazing .. I almost understand parts of it haha!

All I have currently is a photoshop mock up .. It's based on idea which the Wife's 10 year old daughter and I are 'collaborating' on

The final elements will all be hand drawn and have a more fun feel to them .. We hope!

![mock up (test)](https://user-images.githubusercontent.com/85754684/121729029-ea2be980-cae5-11eb-8b34-65e6bce21c0d.jpg)

Background / Character Shape / Eyes / Mouth / Arms / Legs / 'Other Stuff' /

**Reply by DaddyKooL18** [reply]
Each element would be .png

That way it has the transparency and every layer is the same size, so no need to position element :)

![mouth](https://user-images.githubusercontent.com/85754684/121729791-c4531480-cae6-11eb-9007-f17e2b9a64d0.png)

**Reply by DaddyKooL18** [reply]
transparent, not white like its showing there for some reason :)

**Comment by GeeMack** [comment]
Using ImageMagick v7 you can pretty easily select a random image from a group as long as all the images fit in memory. If all the images are the same dimensions and created as transparencies, a single IM command like this could be used to assemble your entire face using randomly selected components every time. This example is in Windows syntax...

    magick ( bg1.png bg2.png bg3.png -duplicate 1,%[fx:floor(rand()*n)] -delete 0--2 ) ^
       ( face1.png face2.png face3.png face4.png -duplicate 1,%[fx:floor(rand()*n)] -delete 0--2 ) ^
       ( eyes1.png eyes2.png eyes3.png -duplicate 1,%[fx:floor(rand()*n)] -delete 0--2 ) ^
       -flatten random_face.png

Inside the first set of parentheses it reads in three background images, duplicates a single random choice from the three, and deletes the three originally read in. In the next set of parentheses it reads in four face images, again selects and duplicates a random choice of the four, and deletes the ones originally read in. The third set of parentheses handles the eye images the same way.

After the parentheses you're left with a randomly chosen background, a randomly chosen face, and a randomly chosen set of eyes, in that order. Simply flatten them into a single image and write the output.

The command should work with any number of input images in each section as long as your system can handle the memory. The input images could be read in as wildcard lists...

    magick ( D:\Project\Eyes\*.png ...

Or the image file names can be read in from text lists by preceding the name of the text files with a "@" like this...

    magick ( @list_of_faces.txt ...

Any paths or file names with spaces must be in double quotes.

Add more components to the faces by adding additional sets of parentheses.

**Reply by DaddyKooL18** [reply]
Good Morning GeeMack :)

Have had a little test with that (putting appropriate files in folders and running this command using your commands) ...

magick ( images\bg\*.png -duplicate 1,%[fx:floor(rand()*n)] -delete 0--2 ) ^
   ( images\egg\*.png -duplicate 1,%[fx:floor(rand()*n)] -delete 0--2 ) ^
   ( images\eyes\*.png -duplicate 1,%[fx:floor(rand()*n)] -delete 0--2 ) ^
   ( images\mouth\*.png -duplicate 1,%[fx:floor(rand()*n)] -delete 0--2 ) ^
   -flatten random_face.png

It kicks back this result ...

C:\Program Files\ImageMagick-7.0.11-Q16-HDRI>magick ( images\bg\*.png -duplicate 1,floor(rand()*n)] -delete 0--2 )    ( images\egg\*.png -duplicate 1,floor(rand()*n)] -delete 0--2 )    ( images\eyes\*.png -duplicate 1,floor(rand()*n)] -delete 0--2 )    ( images\mouth\*.png -duplicate 1,floor(rand()*n)] -delete 0--2 )    -flatten random_face.png
magick: invalid argument for option '-duplicate' '1,floor(rand()*n)]' at CLI arg 3 @ error/operation.c/CLIListOperatorImages/4164.


Also ... could your idea have the output to name sequentially ... ie. eggface001.png , eggfacee002.png etc. please?

and is it possible for it to auto-repeat the script for example 100 times?

Many Thanks :)

**Reply by DaddyKooL18** [reply]
also I do have a \ after each dir before the *.png  .... for some reason its removing that when i paste in here ... same with the resulting window, the path is there with the wildcarded .png

**Reply by GeeMack** [reply]
The first "invalid argument" issue looks like you'll have to double all the percent signs "%%" if you're running it from a BAT script.

**Reply by DaddyKooL18** [reply]
Brilliant! 

Thank you, that has the random face outputted :)

Would you know how to get say 100 automatically sequentially generated please? :)

**Reply by DaddyKooL18** [reply]
Here is where I am upto ...

almost there!  ... Only issue is it overwrites my \images\output\egghead001.png  each time, rather than creating the next inline ... am sure it is something really simple missing / i have incorrect :)


![197371411_2888867534712414_565911158495240086_n](https://user-images.githubusercontent.com/85754684/121767692-c0a8a780-cb51-11eb-9630-e8f22a349ca4.jpg)

**Reply by DaddyKooL18** [reply]
.. ohhh and i still can't work out how to approach this batch with invoking 'run x 100' :D

**Reply by GeeMack** [reply]
There aren't very good ways to handle leading zeros in either Windows CMD or ImageMagick, so a script to output sequential file names might look a little clumsy. This is a Windows BAT script that should run the ImageMagick command 100 times and output files with random combinations of background, face, and eyes, and named "image001.png" through "image100.png".

    @echo off
    setlocal EnableDelayedExpansion
    
    set SEQ=1000
    for /l %%n in ( 1 1 100 ) do (
       set /A SEQ=!SEQ!+1
       magick ^( bg1.png bg2.png bg3.png -duplicate 1,"%%[fx:floor(rand()*n)]" -delete 0--2 ^) ^
          ^( face1.png face2.png face3.png -duplicate 1,"%%[fx:floor(rand()*n)]" -delete 0--2 ^) ^
          ^( eyes1.png eyes2.png eyes3.png -duplicate 1,"%%[fx:floor(rand()*n)]" -delete 0--2 ^) ^
          -flatten image!SEQ:~-3!.png
    )
    
    exit /b

I've tested parts of this but not the whole thing, so it should work but no promises.

**Reply by DaddyKooL18** [reply]
-flatten image!SEQ:~-3!.png    << it seems to have an issue with this line ?

Thanks again! :)

**Reply by DaddyKooL18** [reply]
'(' is not recognized as an internal or external command,
operable program or batch file.
'-flatten' is not recognized as an internal or external command,
operable program or batch file.
magick: unbalanced parenthesis `(end of cli)' at CLI arg 7 @ error/magick-cli.c/ProcessCommandOptions/505.

**Reply by GeeMack** [reply]
Oh there are some continued-line carets I left out. I'll edit the command to fix that. ... It should work now.

**Reply by DaddyKooL18** [reply]
LEGEND!!!

Thank you so much! .. works an absolute treat! :)
